feat: Adds the ECharts Histogram chart

michael-s-molina opened the pull request:
### SUMMARY
This PR creates the ECharts Histogram chart which is used to display the distribution of a dataset by representing the frequency or count of values within different ranges or bins. It helps visualize patterns, clusters, and outliers in the data and provides insights into its shape, central tendency, and spread.

This plugin uses [numpy.histogram](https://numpy.org/doc/stable/reference/generated/numpy.histogram.html#numpy.histogram) to compute the distribution on the server side. It supports a single numeric column and multiple dimensions. Aggregations are not supported.

The legacy Histogram is one of the plugins that are currently preventing us from [upgrading to React 17](https://github.com/orgs/apache/projects/345?pane=issue&itemId=59020521). With this PR, we should be able to create a migration script for legacy charts and remove another blocker for the React upgrade. 

I added the minimum required controls to allow the migration but many more controls can be added in the future to improve the chart's customizability.

Fixes https://github.com/apache/superset/issues/24064
Fixes https://github.com/apache/superset/issues/23066

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<img width="1090" alt="Screenshot 2024-06-03 at 13 15 40" src="https://github.com/apache/superset/assets/70410625/b016aedf-1192-48a3-93d9-5f2b2282ba25">
<img width="1764" alt="Screenshot 2024-06-03 at 15 54 54" src="https://github.com/apache/superset/assets/70410625/a040fb36-8626-4cf1-91a1-217d060c971e">
<img width="1760" alt="Screenshot 2024-06-03 at 15 55 33" src="https://github.com/apache/superset/assets/70410625/c1300df0-ffc7-4af8-8cb1-c600b0ac8b7b">

### TESTING INSTRUCTIONS
Check all features.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
feat: Adds the ECharts Histogram plugin

michael-s-molina commented on superset-frontend/src/explore/components/ExploreViewContainer/index.jsx:
```diff
@@ -460,15 +460,21 @@ function ExploreViewContainer(props) {
-          ),
-      );
+      const { controls } = props;
+      const changedControlKeys = Object.keys(controls).filter(key => {
```
This change fixes a problem with the algorithm that detect a stale chart. Previously, if the values being compared were not objects, we had wrong results. For example, if a user changed the row limit, which is a number, the banner indicating a staled chart was not shown. This should significantly improve user experience when changing control values.

michael-s-molina committed:
Adds cumulative and normalize options

michael-s-molina committed:
Changes column control

rusackas commented:
<img src="https://media0.giphy.com/media/DPznISmq0hLRQqG71g/giphy.gif"/>

mistercrunch commented:
![200w](https://github.com/apache/superset/assets/487433/df6c7dda-59cc-456e-9fd8-e8c858f7bb9f)

michael-s-molina committed:
Adds tests

michael-s-molina commented:
/testenv up

michael-s-molina committed:
Fixes small bug

michael-s-molina commented:
/testenv up

kgabryje commented on superset-frontend/plugins/plugin-chart-echarts/src/Histogram/controlPanel.tsx:
```diff
@@ -0,0 +1,131 @@
+            config: {
+              type: 'CheckboxControl',
+              label: t('Normalize'),
+              description: t('Whether to normalize the histogram'),
```
I wonder if such descriptions are valuable - it feels a bit repetitive with the label. Maybe we should either skip the description, or (preferably) describe how normalization works, or provide a link to a resource that explains normalization?

michael-s-molina replied:
That's a good point. I'll see if I can generate some helpful tooltips using GPT.

kgabryje commented on superset-frontend/plugins/plugin-chart-echarts/src/Histogram/controlPanel.tsx:
```diff
@@ -0,0 +1,131 @@
+            config: {
+              type: 'CheckboxControl',
+              label: t('Cumulative'),
+              description: t('Whether to make the histogram cumulative'),
```
Same comment as above

kgabryje commented on superset-frontend/plugins/plugin-chart-echarts/src/Histogram/controlPanel.tsx:
```diff
@@ -0,0 +1,131 @@
+              label: t('X Axis Title'),
+              renderTrigger: true,
+              default: '',
+              description: t('Changing this control takes effect instantly'),
```
I think that tooltip is already displayed when hovering the `renderTrigger: true` icon?

kgabryje commented on superset-frontend/plugins/plugin-chart-echarts/src/Histogram/controlPanel.tsx:
```diff
@@ -0,0 +1,131 @@
+              label: t('Y Axis Title'),
+              renderTrigger: true,
+              default: '',
+              description: t('Changing this control takes effect instantly'),
```
here too

kgabryje commented on superset-frontend/plugins/plugin-chart-echarts/src/Histogram/transformProps.ts:
```diff
@@ -0,0 +1,203 @@
+  } = formData;
+  const { data } = queriesData[0];
+  const colorFn = CategoricalColorNamespace.getScale(colorScheme);
+  const formatter = getNumberFormatter(
```
Should we also support currency formatter here?

michael-s-molina replied:
Let me give you some context to answer this question. Now that we're migrating legacy plugins, it becomes clear that we have a lot of duplicated code and many inconsistencies between ECharts plugins regarding the controls that are available. I'm planning to tackle this horizontally, making sure that features, such as currency formatting, are available consistently among the plugins. For that reason, for this PR:

> I added the minimum required controls to allow the migration but many more controls can be added in the future to improve the chart's customizability.

kgabryje replied:
Sorry, missed that part of the pr description. Makes sense!

michael-s-molina replied:
No problem! You correctly spotted the missing control.

kgabryje commented on superset-frontend/plugins/plugin-chart-echarts/src/Histogram/transformProps.ts:
```diff
@@ -0,0 +1,203 @@
+  const percentFormatter = getPercentFormatter(NumberFormats.PERCENT_2_POINT);
+  const groupbySet = new Set(groupby);
+  const xAxisData: string[] = Object.keys(data[0]).filter(
+    key => groupbySet.has(key) === false,
```
```suggestion
    key => !groupbySet.has(key),
```

kgabryje commented on superset-frontend/packages/superset-ui-chart-controls/src/utils/columnChoices.ts:
```diff
@@ -16,24 +16,29 @@
+    const { columns } = datasource;
+    const filteredColumns = (
+      columns as { type_generic?: GenericDataType }[]
+    ).filter(col => (type !== undefined ? col.type_generic === type : true));
```
```suggestion
    ).filter(col => (type === undefined || col.type_generic === type));
```

michael-s-molina committed:
Addresses comments

michael-s-molina commented:
Thanks for the review @kgabryje! I think I addressed all your comments, including adding helpful tooltips. Thanks for pointing that out! It was a copy/paste from the legacy chart.

michael-s-molina commented:
/testenv up

kgabryje commented:
@michael-s-molina It looks like normalised histogram doesn't reach 1 for tallest bar - 1st screenshot is the new plugin, 2nd is the old one

<img width="1073" alt="image" src="https://github.com/apache/superset/assets/15073128/269f3ec6-d538-44ae-b381-6969ac3fc2dc">

<img width="1327" alt="image" src="https://github.com/apache/superset/assets/15073128/82a07c98-36ac-4c2a-9b13-55285c9b905e">

kgabryje commented:
1 more comment - why do we have a select component for num column instead of drag&drop? We could limit the accepted column by type in dnd

michael-s-molina committed:
Changes normalize calculation

michael-s-molina commented:
> @michael-s-molina It looks like normalised histogram doesn't reach 1 for tallest bar - 1st screenshot is the new plugin, 2nd is the old one

@kgabryje I changed the normalize calculation and updated the tooltip. It should be correct now. Keep in mind that it will not match the legacy version because that was wrong too 😅 .

kgabryje reviewed (approved):
LGTM, thanks for creating a proper histogram plugin!

michael-s-molina committed:
Changes column control to use dnd

michael-s-molina commented:
> 1 more comment - why do we have a select component for num column instead of drag&drop? We could limit the accepted column by type in dnd

@kgabryje I changed the control to use drag-and-drop as you suggested.

michael-s-molina committed:
Fixes type error

michael-s-molina committed:
Add another test

michael-s-molina committed:
Fixes import
